### DISCUSSION TITLE
Listeners do not seem to trigger normally while testing

### DISCUSSION THREAD

**Discussion by JHWelch** [livewire/livewire#3383] [7. Bugs (archived)] [answered]
Not sure if this is simply a misunderstanding on my part, but I found that listeners do not seem to fire normally from events emitted in functions while testing a Livewire component.

I was able to recreate this exact problem in [this new repository](https://github.com/JHWelch/testing-livewire-emits)

If I have this simple Livewire component
```php
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class EmitTest extends Component
{
    public $variable = 'x';

    protected $listeners = [
        'emitted-event' => 'changeVariable',
    ];

    public function emitEvent()
    {
        $this->emit('emitted-event');
    }

    public function changeVariable()
    {
        $this->variable = 'successfully emitted';
    }
}
```

with this template:
```html
<div>
    <button wire:click="emitEvent">Emit to object</button>
    <span>{{ $variable }}</span>
</div>
```

And try and test it like this
```php
/** @test */
public function event_is_emitted_but_page_is_not_updated()
{
    $component = Livewire::test(EmitTest::class)
        ->call('emitEvent');

    $component->assertEmitted('emitted-event');
    // This will fail.
    $component->assertSee('successfully emitted');
}
```

The test will not see any changes to the page, even though it **does** see the event emitted. Testing this page manually shows it working as well.

In this test I show the workaround I have ended up doing in my own tests, where I first check to see if the event is emitted, and then manually emit it with the `TestableLivewire` class. This does work, and triggers the listener as expected.
```php
/** @test */
public function you_can_hack_this_by_checking_for_emit_and_manually_emitting()
{
    $component = Livewire::test(EmitTest::class)
        ->call('emitEvent');

    // Check to see if it is emitted, and then emit.
    $component->assertEmitted('emitted-event');
    $component->emit('emitted-event');

    // This works
    $component->assertSee('successfully emitted');
}
```

If listeners are not supposed to work normally in tests, it would be great if that was explained in the documentation somewhere. If I missed it, I apologize.

**Comment by joshhanley** [comment, marked as answer]
@JHWelch yeah listeners are actually fired on the frontend using javascript, but the tests are php only tests, so they won't fire.

The way you have structured your test seems good to me.

I'm sure a PR to the docs to clarify would be welcome.

Hope this helps!

**Reply by joshhanley** [reply]
I would potentially even make it two tests, one to ensure the event was fired, and another to ensure it reacts to the event as expected.

**Reply by JHWelch** [reply]
That makes sense! Didn't think about how the events were actually being fired. I agree that the best structure then would be to have one test up to the event being fired, asserting it was fired and then another test with a fake event testing the listener.

Though this make it harder to test a full feature depending. Wonder if there is a good way to fake the event in PHP in the TestableLivewire class, though doesn't seem a common problem.
